### PR TITLE
[FLINK-2853] [tests] Apply JMH on MutableHashTablePerformanceBenchmark class.

### PR DESCRIPTION
**Pull request by gallenvara**
JMH is a Java harness for building, running, and analysing nano/micro/milli/macro benchmarks.Use JMH to replace the old micro benchmarks method in order to get much more accurate results.Modify the` MutableHashTablePerformanceBenchmark` class and move it to `flink-benchmark` module.

### CODE REVIEW

**Comment by gallenvara** [comment]
JMH result (We can choose the benchmark mode with througout, averagetime,sampletime and so on.Also we can set the warmup and measurement iteration number according to test need.) :
jmgtestresult

File: flink-runtime/pom.xml [hunk 1]
```diff
@@ -210,7 +210,12 @@ under the License.
 			<version>${curator.version}</version>
 			<scope>test</scope>
 		</dependency>
-	</dependencies>
+        <dependency>
```
Comment (fhueske) [inline flink-runtime/pom.xml]: Why do you add a dependency to itself (this is the pom file of `flink-runtime`)?

File: pom.xml [hunk 2]
```diff
@@ -163,6 +163,16 @@ under the License.
 			<type>jar</type>
 			<scope>test</scope>
 		</dependency>
+        <dependency>
```
Comment (fhueske) [inline pom.xml]: I don't think it is necessary to add these dependency to the root pom. We added a dedicated Maven module for micro benchmarks (`flink-benchmark`). This is where the code and the dependencies should go, IMO.

File: flink-benchmark/pom.xml [hunk 3]
```diff
@@ -57,6 +57,13 @@ under the License.
       <version>${jmh.version}</version>
       <scope>provided</scope>
     </dependency>
+      <dependency>
+          <groupId>org.apache.flink</groupId>
+          <artifactId>flink-runtime</artifactId>
+          <version>0.10-SNAPSHOT</version>
```
Comment (fhueske) [inline flink-benchmark/pom.xml]: We don't hardcode the version for flink dependencies. Please use `<version>${project.version}</version>` instead.

File: flink-benchmark/src/test/java/org/apache/flink/runtime/operators/hash/MutableHashTablePerformanceBenchmark.java [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+import static org.junit.Assert.fail;
+
+@State(Scope.Thread)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+public class MutableHashTablePerformanceBenchmark {
```
Comment (fhueske) [inline flink-benchmark/src/test/java/org/apache/flink/runtime/operators/hash/MutableHashTablePerformanceBenchmark.java]: Can you move this to the `flink-benchmark` module?

Comment (gallenvara) [inline flink-benchmark/src/test/java/org/apache/flink/runtime/operators/hash/MutableHashTablePerformanceBenchmark.java]: This class is already under the `flink-benchmark` module.The package name is confused(same as `flink-runtime` module in order to keep the structure), i will change the package name.

File: pom.xml (hunk 2 above)
Comment (gallenvara) [inline pom.xml]: Yes,that's right.I have done some work in `flink-runtime` module and `flink-core` before,so i add JMH dependency to the root pom in oder to apply to all modules.I will modify the pom and move the dependency to pom of `flink-benchmark`.

File: flink-benchmark/pom.xml (hunk 3 above)
Comment (gallenvara) [inline flink-benchmark/pom.xml]: Thanks for your review.I will modify the pom as you have suggested.

File: flink-runtime/pom.xml [hunk 5]
```diff
@@ -210,7 +210,7 @@ under the License.
 			<version>${curator.version}</version>
 			<scope>test</scope>
 		</dependency>
-	</dependencies>
+    </dependencies>
```
Comment (ChengXiangLi) [inline flink-runtime/pom.xml]: Should not change the format for no reason.

File: flink-benchmark/src/test/java/org/apache/flink/benchmark/runtime/operates/hash/MutableHashTablePerformanceBenchmark.java [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+		// create a probe input with 5000000 records with key spread between [0 -- 1000000] with distance of 1 for nearby records.
+		int probeSize = 5000000;
+		int probeStep = 1;
+		int probeScope = buildSize;
+
+		int expectedResult = 500000;
+
+		this.hybridHashJoin(buildSize, buildStep, buildScope, probeSize, probeStep, probeScope, expectedResult, true);
+
+		System.out.println("HybridHashJoin2:");
```
Comment (ChengXiangLi) [inline flink-benchmark/src/test/java/org/apache/flink/benchmark/runtime/operates/hash/MutableHashTablePerformanceBenchmark.java]: This should be "HybridHashJoin1", right?

File: flink-runtime/pom.xml (hunk 5 above)
Comment (gallenvara) [inline flink-runtime/pom.xml]: Thanks for your review.I will correct it.

File: flink-benchmark/src/test/java/org/apache/flink/benchmark/runtime/operates/hash/MutableHashTablePerformanceBenchmark.java (hunk 6 above)
Comment (gallenvara) [inline flink-benchmark/src/test/java/org/apache/flink/benchmark/runtime/operates/hash/MutableHashTablePerformanceBenchmark.java]: Yes,that's right.I have submitted a new PR for correcting these problems.

**Comment by ChengXiangLi** [comment]
LGTM, +1

File: flink-benchmark/pom.xml [hunk 7]
```diff
@@ -57,6 +57,13 @@ under the License.
       <version>${jmh.version}</version>
       <scope>provided</scope>
     </dependency>
+      <dependency>
```
Comment (fhueske) [inline flink-benchmark/pom.xml]: indention is off by 2 chars

**Comment by fhueske** [comment]
The `MutableHashTablePerformanceBenchmark` can be removed from `flink-runtime` if it is added to `flink-benchmark`.

Otherwise looks good.

**Comment by fhueske** [comment]
Looks good to merge.

**Comment by fhueske** [comment]
I'm going to merge this PR.

**Pull request closed without merging**